src/Connect.js

```diff
@@ -1,68 +1,441 @@
+    // TODO remove defaults, fix import
+    const subProvider = new provider.default({
+      requestAddress: () => {
+        this.requestAddress('addressReqProvider')
+        return this.onResponse('addressReqProvider').then(payload => {
+          this.setDID(payload.res.address)
+          return this.address
+        })
+
+      },
+      sendTransaction: (txObj) => {
+        delete txObj['from']
+        this.sendTransaction(txObj, 'txReqProvider')
+        return this.onResponse('txReqProvider').then(payload => payload.res)
+      },
+      provider: this.provider,
+      networkId: this.network.id
+    })
+    if (this.address) subProvider.setAccount(this.address)
+    return subProvider
+  }
+
+// TODO where to return MNID and where to return address, should this be named differently, will return entire response obj now, not just address
+// TODO requestID? requestAddress? return mnid, address, did in response??
+// TODO add account option, param
+ /**
+  *  Creates a request for only the address/id of the uPort identity.
+  *
+  *  @example
+  *  connect.requestAddress()
+  *
+  *  connect.onResponse('addressReq').then(res => {
+  *    const id = res.res
+  *  })
+  *
+  *  @param    {String}    [id='addressReq']    string to identify request, later used to get response
+  */
+  requestAddress (id='addressReq') {
+    this.credentials.requestDisclosure({callbackUrl: this.genCallback()})
+                    .then(jwt => this.request(jwt, id))
   }
 
+ // TODO offer listener and single resolve? or other both for this funct, by allowing optional cb instead
  /**
-  *  Instantiates and returns a web3 object wrapped with uPort functionality. For
-  *  more details see uportSubprovider and getProvider in connectCore.
+  *  Get response by id of earlier request, returns promise which resolves when first reponse with given id is available. Listen instead, if looking for multiple responses of same id.
   *
-  *  @return          {web3}    A uPort web3 object
+  *  @param    {String}    id             id of request you are waiting for a response for
+  *  @return   {Promise<Object, Error>}   promise resolves once valid response for given id is avaiable, otherwise rejects with error
   */
-  getWeb3 () {
-    const provider = this.getProvider()
-    const web3 = new Web3()
-    web3.setProvider(provider)
-    // Work around to issue with web3 requiring a from parameter. This isn't actually used.
-    web3.eth.defaultAccount = '0xB42E70a3c6dd57003f4bFe7B06E370d21CDA8087'
-    return web3
+  onResponse(id) {
+    const parseResponse = (payload) => {
+      if (payload.error) return Promise.reject(Object.assign({id}, payload))
+      if (message.util.isJWT(payload.res)) {
+        const jwt = payload.res
+        const decoded = decodeJWT(jwt)
+        if (decoded.payload.claim){
+          return Promise.resolve(Object.assign({id}, payload))
+        }
+        return this.verifyResponse(jwt).then(res => {
+            this.setDID(res.address)
```

Comment:
And the `nad`, that should be saved in state as well and used for ethereum interactions